networkstatsquerier/main.go

```diff
@@ -27,63 +28,108 @@ var totalUptimeChecksCounterVec *prometheus.CounterVec
-	// Closing context if HTTP request and response processing is not completed within 30 seconds.
-	// This includes dropping the request connection if there's no reply within 30 seconds.
-	ctx, cancelFunc := context.WithDeadline(ctx, time.Now().Add(30*time.Second))
-	defer cancelFunc()
+func getHTTPRequestTiming(message []byte, args *map[string]interface{}) error {
+	// Get arguments
+	wkrs, ok := (*args)["workers"].(*workers.Workers)
+	if !ok {
+		return fmt.Errorf("unable to cast Workers from arguments")
+	}
+	ctx, ok := (*args)["ctx"].(context.Context)
+	if !ok {
+		return fmt.Errorf("unable to cast context from arguments")
+	}
+	numWorkers, ok := (*args)["numWorkers"].(int)
+	if !ok {
+		return fmt.Errorf("unable to cast numWorkers to int from arguments")
+	}
+	errs, ok := (*args)["errs"].(chan error)
+	if !ok {
+		return fmt.Errorf("unable to cast errs to chan error from arguments")
+	}
+	jobDuration, ok := (*args)["jobDuration"].(time.Duration)
+	if !ok {
+		return fmt.Errorf("unable to cast jobDuration to time.Duration from arguments")
+	}
 
-	var resp, responseTime, err = querier.GetResponseAndTiming(ctx, urlString)
+	// Handle the start message that is sent before the endpoints and the stop message that is sent at the end
+	if string(message) == "start" {
+		err := wkrs.Start(ctx, numWorkers, errs)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	if string(message) == "stop" {
+		err := wkrs.Stop()
+		if err != nil {
+			return fmt.Errorf("error stopping queue workers: %s", err.Error())
+		}
+		return nil
+	}
 
+	urlString := string(message)
+	// Specifically query the FHIR endpoint metadata
+	metadataURL, err := url.Parse(urlString)
 	if err != nil {
-		log.WithFields(log.Fields{"organization": organizationName, "url": urlString}).Warn("Error getting response charactaristics for endpoint.", err.Error())
-	} else {
-		responseTimeGaugeVec.WithLabelValues(organizationName).Set(responseTime)
+		return fmt.Errorf("Endpoint URL Parsing Error: %s", err.Error())
+	}
 
-		if resp != nil && resp.StatusCode != http.StatusOK {
-			totalFailedUptimeChecksCounterVec.WithLabelValues(organizationName).Inc()
-		}
-		if resp != nil {
-			httpCodesGaugeVec.WithLabelValues(organizationName).Set(float64(resp.StatusCode))
-		}
-		totalUptimeChecksCounterVec.WithLabelValues(organizationName).Inc()
+	metadataURL.Path = path.Join(metadataURL.Path, "metadata")
+
+	jobArgs := make(map[string]interface{})
+	jobArgs["urlString"] = metadataURL.String()
+	jobArgs["respTime"] = responseTimeGaugeVec
+	jobArgs["failUptime"] = totalFailedUptimeChecksCounterVec
+	jobArgs["httpCodes"] = httpCodesGaugeVec
+	jobArgs["totalUptime"] = totalUptimeChecksCounterVec
+
+	job := workers.Job{
+		Context:     ctx,
+		Duration:    jobDuration,
+		Handler:     (querier.GetResponseAndTiming),
+		HandlerArgs: &jobArgs,
+	}
+
+	err = wkrs.Add(&job)
+	if err != nil {
+		return fmt.Errorf("error adding job to workers: %s", err.Error())
 	}
+
+	return nil
```

Comment:
this is a fair amount of copied code based on what was done in capability querier, which always makes me a bit uncomfortable.
if it's easy to do, consider refactoring. if not, it can be a future cleanup task.

Reply:
I started looking into this but it seemed like more effort than it was worth at the current moment, especially if there's the possibility that the networkstatsquerier and the capability querier get combined in the future.